File: src/main/java/io/github/aphelium/modules/brokers/useCases/createBroker/CreateBrokerUseCase.java
package io.github.aphelium.modules.brokers.useCases.createBroker;


import io.github.aphelium.modules.brokers.dtos.CreateBrokerDTO;
import io.github.aphelium.modules.brokers.entities.Broker;
import io.github.aphelium.modules.brokers.exceptions.BrokerAlreadyExistsException;
import io.github.aphelium.modules.brokers.repositories.BrokerRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.time.LocalDateTime;
import java.time.ZoneId;


@Service
public class CreateBrokerUseCase {

    final BrokerRepository brokerRepository;

    public CreateBrokerUseCase(BrokerRepository brokerRepository) {
        this.brokerRepository = brokerRepository;
    }

    @Transactional
    public Broker execute(CreateBrokerDTO dto) throws BrokerAlreadyExistsException {

        Broker broker = new Broker();

        // Check if the broker already exists
        if(brokerRepository.findByName(dto.getName()).isPresent()) {
            throw new BrokerAlreadyExistsException();
        }


        BeanUtils.copyProperties(dto, broker);

        broker.setCreatedAt(LocalDateTime.now(ZoneId.of("UTC")));
        broker.setUpdatedAt(LocalDateTime.now(ZoneId.of("UTC")));

        brokerRepository.save(broker);

        return broker;

    }

}

File: src/main/kotlin/io/github/aphelium/modules/brokers/useCases/createBroker/CreateBrokerUseCase.kt
package io.github.aphelium.modules.brokers.useCases.createBroker

import io.github.aphelium.modules.brokers.dtos.CreateBrokerDTO
import io.github.aphelium.modules.brokers.entities.Broker
import io.github.aphelium.modules.brokers.exceptions.BrokerAlreadyExistsException
import io.github.aphelium.modules.brokers.repositories.BrokerRepository
import org.springframework.beans.BeanUtils
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.time.ZoneId
import javax.transaction.Transactional

@Service
class CreateBrokerUseCase(val brokerRepository: BrokerRepository) {
    @Transactional
    @Throws(BrokerAlreadyExistsException::class)
    fun execute(dto: CreateBrokerDTO?): Broker {
        val broker = Broker()

        // Check if the broker already exists
        if (brokerRepository.findByName(dto!!.name).isPresent) {
            throw BrokerAlreadyExistsException()
        }
        BeanUtils.copyProperties(dto, broker)
        broker.createdAt = LocalDateTime.now(ZoneId.of("UTC"))
        broker.updatedAt = LocalDateTime.now(ZoneId.of("UTC"))
        brokerRepository.save(broker)
        return broker
    }
}